### PR TITLE
refactor: refactor static regex out of for loops

### LINKED ISSUE #12071: Refactor static regEx out of for loops
### Describe the proposed change(s).

We should not be creating static or repeated `regEx()` patterns repeatedly inside for loops. Such patterns which are executed repeatedly inside for loops should be moved up in scope to the point where they are created once per pattern.

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes:

Reverted #12071  flags from regex as it is no longer relevant. 
Refactored duplicate regex.

## Context:

Closes #12071 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/migration.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import { dequal } from 'dequal';
 import { logger } from '../logger';
 import { clone } from '../util/clone';
 import { regEx } from '../util/regex';
+import { fixShortHours } from '../workers/branch/schedule';
```
Comment (viceice) [inline lib/config/migration.ts:7]: Nope. Config can't have a dependency on workers.

So this function needs to be moved to config or utils.

File: lib/datasource/artifactory/index.ts [hunk 2]
```diff
@@ -109,6 +109,6 @@ export class ArtifactoryDatasource extends Datasource {
   }
 
   private static parseReleaseTimestamp(rawText: string): string {
-    return rawText.trim().replace(regEx(/ ?-$/), '') + 'Z'; // TODO #12071
+    return rawText.trim().replace(regEx(/ ?-$/), '');
```
Comment (viceice) [inline lib/datasource/artifactory/index.ts:112]: Wrong

**Review by viceice — COMMENTED** [review]
More later

File: lib/config/migration.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/config/migration.ts:7]: Moved it to `util/url.ts`

File: lib/util/url.ts [hunk 3]
```diff
@@ -79,3 +79,7 @@ export function parseUrl(url: string): URL | null {
 export function createURLFromHostOrURL(url: string): URL | null {
   return parseUrl(url) ?? parseUrl(`https://${url}`);
 }
+
+export function fixShortHours(input: string): string {
```
Comment (viceice) [inline lib/util/url.ts:83]: Definitely wrong file. 🤔

Move to `./date.ts`

File: lib/config/migration.ts [hunk 4]
```diff
@@ -3,6 +3,7 @@ import is from '@sindresorhus/is';
 import { dequal } from 'dequal';
 import { logger } from '../logger';
 import { clone } from '../util/clone';
+import { fixShortHours } from '../util/date';
```
Comment (rarkins) [inline lib/config/migration.ts:6]: @viceice are you sure this shouldn't remain part of migration?

File: lib/config/migration.ts [hunk 5]
```diff
@@ -19,7 +20,9 @@ import { mergeChildConfig } from './utils';
 const options = getOptions();
 
 let optionTypes: Record<string, RenovateOptions['type']>;
-
+const afterBeforeRe = regEx(
+  /^(.*?)(after|before) (.*?) and (after|before) (.*?)( |$)(.*)/
+);
```
Comment (rarkins) [inline lib/config/migration.ts:25]: @viceice do you think we need to push these regEx up to module scope or can they be located more close to where they're used?

Comment (RahulGautamSingh) [reply lib/config/migration.ts:25]: @viceice Could you please review this once?

Comment (viceice) [reply lib/config/migration.ts:25]: I think this can now be moved more close to usage now

File: lib/config/migration.ts (hunk 4 above)
Comment (rarkins) [reply lib/config/migration.ts:6]: sure, but that's maybe a separate PR?

Comment (rarkins) [reply lib/config/migration.ts:6]: @RahulGautamSingh please skip this one for now, it should be a separate issue (to move it to migration)

**Comment by RahulGautamSingh** [comment]
@viceice @rarkins Can you review it, please? It's been stale for quite a while now.

File: lib/util/date.ts [hunk 6]
```diff
@@ -10,3 +12,7 @@ export function getElapsedDays(timestamp: string): number {
 export function getElapsedMinutes(date: Date): number {
   return Math.floor((new Date().getTime() - date.getTime()) / ONE_MINUTE_MS);
 }
+
+export function fixShortHours(input: string): string {
+  return input.replace(regEx(/( \d?\d)((a|p)m)/g), '$1:00$2');
```
Comment (rarkins) [inline lib/util/date.ts:17]: @viceice I'm not sure about this relocation.. it's very later-specific, and I think belongs in migration. Or do you think there's a better reason for it being here?

**Review by rarkins — COMMENTED** [review]
What if in the next PR we changed our `regEx` from:

```
export function regEx(
  pattern: string | RegExp,
  flags?: string | undefined,
  useCache = true
): RegExp {
```

to

```
export function regEx(
  pattern: string,
  flags?: string | undefined,
  useCache = true
): RegExp {
```

We still have a lot of `regEx(/.../)` inside of loops and although we cache the `re2` result, we're still dynamically creating a `RegExp` each loop due to the `/` constructor.

**Comment by RahulGautamSingh** [comment]
> What if in the next PR we changed our `regEx` from:
> 
> ```
> export function regEx(
>   pattern: string | RegExp,
>   flags?: string | undefined,
>   useCache = true
> ): RegExp {
> ```
> 
> to
> 
> ```
> export function regEx(
>   pattern: string,
>   flags?: string | undefined,
>   useCache = true
> ): RegExp {
> ```
> 
> We still have a lot of `regEx(/.../)` inside of loops and although we cache the `re2` result, we're still dynamically creating a `RegExp` each loop due to the `/` constructor.

For the case of `regEx(/.../)` inside of loops, the RegExp object that is passed to the `regEx` function is created only once per page. So, we don't need to worry about it being created multiple times.
Source:-
```
Regular expression literals provide compilation of the regular expression when the script is evaluated. When the regular expression will remain constant, use this for better performance.
```
[Link](http://www.websina.com/bugzero/kb/regexp.html)

File: lib/util/date.ts (hunk 6 above)
Comment (viceice) [reply lib/util/date.ts:17]: Yeah, let's move to migration code. Hopefully we can remove the call from worker code soon

Comment (rarkins) [reply lib/util/date.ts:17]: @RahulGautamSingh could you make this change?

Comment (RahulGautamSingh) [reply lib/util/date.ts]: @rarkins Done.

**Review by viceice — APPROVED** [review]

**Pull request merged**